File: genesys-characters/src/js/genesys.js
// This module contains definitions for types used in the Genesys CharacterSheets.

/**
 * @readonly
 * @enum {string}
 */
export const CHARACTERISTIC = {
    Brawn: "Brawn",
    Agility: "Agility",
    Intellect: "Intellect",
    Cunning: "Cunning",
    Willpower: "Willpower",
    Presence: "Presence"
 };

 export class Characteristic {
     static Shorten(characteristic) {
        switch (characteristic) {
            case CHARACTERISTIC.Brawn: return "Br";
            case CHARACTERISTIC.Agility: return "Ag";
            case CHARACTERISTIC.Intellect: return "Int";
            case CHARACTERISTIC.Cunning: return "Cun";
            case CHARACTERISTIC.Willpower: return "Will";
            case CHARACTERISTIC.Presence: return "Pr";
        }
        console.warn("tried to shorten '" + characteristic + "' but it wasn't reconised as a characteristic");
        return characteristic;
     }

     static Normalize(characteristic) {

     }
 }
 
 /**
  * @param {CHARACTERISTIC} characteristic 
  */
 export const ShortenCharacteristic = (characteristic) => {
     return Characteristic.Shorten(characteristic);
 }

/**
 * @readonly
 * @enum {string}
 */
export const RANGE = {
    Engaged: "Engaged",
    Short: "Short",
    Medium: "Medium",
    Long: "Long",
    Extreme: "Extreme",
}

/**
 * @typedef {Object} Skill
 * @property {string} name
 * @property {CHARACTERISTIC} characteristic
 * @property {boolean} career
 * @property {number} rank
 */
 export class Skill {
    /**
     * 
     * @param {string} name 
     * @param {CHARACTERISTIC} characteristic
     * @param {boolean} career 
     * @param {number} rank 
     */
    constructor(name, characteristic, career, rank) {
        this.name = name;
        this.characteristic = characteristic;
        this.career = career;
        this.rank = rank;
    }
}

/**
 * @typedef {Object} Favor
 * @property {string} text
 */
export class Favor {
    constructor(text) {
        this.text = text;
    }
}

/**
 * @typedef {Object} Weapon
 * @property {string} name 
 * @property {string} skill 
 * @property {number} damage 
 * @property {bool} uses_brawn
 * @property {number} crit 
 * @property {RANGE} range 
 * @property {string} special 
 * @property {number} encumberance
 * @property {number} hard_points
 * @property {string} description
 */
export class Weapon {
    /**
     * 
     * @param {string} name 
     * @param {string} skill 
     * @param {number} damage 
     * @param {bool} uses_brawn true if brawn score should be added to damage
     * @param {number} crit 
     * @param {RANGE} range 
     * @param {string} special 
     * @param {number} encumberance
     * @param {number} hard_points
     * @param {string} description
     */
    constructor(name, skill, damage, uses_brawn, crit, range, special, encumberance, hard_points, description) {
        this.name = name;
        this.skill = skill;
        this.damage = damage;
        this.uses_brawn = uses_brawn;
        this.crit = crit;
        this.range = range;
        this.special = special;
        this.encumberance = encumberance;
        this.hard_points = hard_points;
        this.description = description;
    }
}

/**
 * @typedef {Object} Item
 * @property {string} name
 * @property {number} quantity
 * @property {string} description
 */
export class Item {
    constructor(name, quantity, description) {
        this.name = name;
        this.quantity = quantity;
        this.description = description;
    }
}

/**
 * @typedef {Object} CriticalInjury
 * @property {number} severity
 * @property {string} result
 */
export class CriticalInjury {
    constructor(severity, result) {
        this.severity = severity;
        this.result = result;
    }
}

/**
 * @typedef {Object} Ability
 * @property {string} name
 * @property {number} rank
 * @property {string} source
 * @property {string} description
 */
export class Ability {
    constructor(name, rank, source, description) {
        this.name = name;
        this.rank = rank;
        this.source = source;
        this.description = description;
    }
}

/**
 * @typedef {Object} Mechanic
 * @property {string} mechanic_type
 * @property {number} magnitude
 * @property {string} description
 */
export class Mechanic {
    constructor(type, magnitude, description) {
        this.mechanic_type = type;
        this.magnitude = magnitude;
        this.description = description;
    }
}

File: genesys-characters/src-tauri/src/genesys.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Default, Clone)]
pub struct Character {
    pub header: Header,
    pub description: Description,
    pub characteristics: Characteristics,
    pub inventory: Vec<Item>,
    pub motivations: Vec<Mechanic>,
    pub obligations: Vec<Mechanic>,
    pub obligation_total: i32,
    pub notes: Vec<Note>,
    pub skills_general: Vec<Skill>,
    pub skills_magic: Vec<Skill>,
    pub skills_combat: Vec<Skill>,
    pub skills_social: Vec<Skill>,
    pub skills_knowledge: Vec<Skill>,
    pub weapons: Vec<Weapon>,
    pub abilities: Vec<Ability>,
    pub critical_injuries: Vec<CriticalInjury>,
    pub favors_given: Vec<Favor>,
    pub favors_owed: Vec<Favor>,
}

#[derive(Serialize, Deserialize, Debug, Default, Clone)]
pub struct Header {
    pub name: String,
    pub player: String,
    pub archetype: String,
    pub career: String,
    pub specializations: String,
    pub xp_available: i32,
    pub xp_total: i32,
}

#[derive(Serialize, Deserialize, Debug, Default, Clone)]
pub struct Description {
    pub age: String,
    pub build: String,
    pub eyes: String,
    pub features: String,
    pub hair: String,
    pub gender: String,
    pub height: String,
}

#[derive(Serialize, Deserialize, Debug, Default, Clone)]
pub struct Characteristics {
    pub brawn: i32,
    pub agility: i32,
    pub intellect: i32,
    pub cunning: i32,
    pub willpower: i32,
    pub presence: i32,
    pub force_rank: i32,
    pub soak: i32,
    pub wounds_theshold: i32,
    pub wounds_current: i32,
    pub strain_threshold: i32,
    pub strain_current: i32,
    pub defense_melee: i32,
    pub defense_ranged: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Mechanic {
    pub mechanic_type: String,
    pub description: String,
    pub magnitude: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Favor {
    pub text: String,
}


#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Item {
    pub quantity: i32,
    pub name: String,
    pub description: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum Characteristic {
    Brawn,
    Agility,
    Intellect,
    Cunning,
    Willpower,
    Presence,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Skill {
    pub name: String,
    pub characteristic: Characteristic,
    pub career: bool,
    pub rank: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
#[serde(default)]
pub struct Note {
    pub note_title: String,
    pub subtitle: String,
    pub body: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Ability {
    pub name: String,
    pub description: String,
    pub rank: i32,
    pub source: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Weapon {
    pub name: String,
    pub description: String,
    pub damage: i32,
    pub critical: i32,
    pub encumbrance: i32,
    pub hard_points: i32,
    pub range: Range,
    pub skill: String,
    pub special: String,
    pub uses_brawn: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum Range {
    Enganged,
    Short,
    Medium,
    Long,
    Extreme,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct CriticalInjury {
    pub severity: i32,
    pub result: String,
}